### ISSUE TITLE
Bitmap images --- extremely slow

### ISSUE THREAD

**Issue by zettam** [xseignard/thermalPrinter#17]
I'm trying to print a bitmap image. the print works but the output is EXTREMELY slow. it waits almost a second between each line - so the paper comes out as slow as a huge plotter.

How can I fix this?

**Comment by zettam** [comment]
the problem is related to ````.drain```` - however I could not find a solution. For some reason ````.drain```` callback is delayed around a second --- which delays the whole print operation a second per line.

**Comment by xseignard** [comment]
Hello, did you tried with other formats?

You could also try to remove the `drain` since it is not mandatory.

This issue may be related to `serialport` package.

Regards

**Comment by zettam** [comment]
Yes I've tried other formats. I think the issue is related to how buffers are being created.
Instead of creating line by line buffers, the code creates bit by bit buffers - which overflows the printer.

Here is a quickfix:
look at this line https://github.com/xseignard/thermalPrinter/blob/master/src/printer.js#L266

instead, using this:
````
commands.push(buf);
````


I may be missing a point, but, the original code creates a buffer per each decimal wheras this one pushes everything into the same line.

This way .drain works okay (or, better).

Let me know if I'm missing a point.

**Comment by zettam** [comment]
The same should be done with everything else, as everything is created with many many buffers. I could not modify the code to do the same with text as I'm not good with buffers..

**Comment by xseignard** [comment]
AFAIK with `push` you can only push a single element to an array, not an array to an array. This explains the `apply` stuff.

See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push#Merging_two_arrays

**Comment by xseignard** [comment]
Be sure to have a good power supply for the printer too, it affects the printing speed

**Comment by zettam** [comment]
You're right abotu the push and apply - but if you try - it solves the slowdown problem. The slowdown is caused by the "drain".

To mention: maybe it's about the processing speeds of raspbeery pi -- as far as my tests go, it's not possible to handle too many drains - maybe that's one of the causes.

However when you do only push (and not apply) it's already being converted to buffer when it's being sent as a command - therefore it's still the same.


At the moment the main issue I'm having is that -- from time to time the images do not get printed correctly. They're partly chopped and slided, or in the middle of a printing session they start to come out as texts.

**Comment by xseignard** [comment]
Yes I have the same problem with images not printed correctly and weird text printed instead (I noticed it on Ubuntu machine and Mac mini). I don't know where it comes from. My guess it that something is badly interpreted on the printer side. We should investigate.

For the `push` vs. `apply` stuff, I don't have any printers (nor time) left to test it so I cannot tell you if using the `push` as you say will affect stuff or not.

I'll let you know

**Comment by zettam** [comment]
Thank you for the replies.
For the "text instead of image" problem, I believe it's something about printers buffer. As far as my tests go, it generally happens when the images have more black then more white - and with a higher potential of happening with bigger/longer pictures.

With a simple logic I believe it's because images with black pixels take longer to print than white ones. In other words, printer can not handle all data bombing and throws out stuff as text.

My solution was, to lower down heat values to increase speed. When the ````heatingTime```` is below 70 I did not have any problems - also surely have to put this as a side note that images are not as crisp/dark as they are with higher values.

**Comment by xseignard** [comment]
Yes I also noticed the same.

Here are the usual values I use:
```js
// printer opts
var opts = {
	maxPrintingDots: 15,
	heatingTime: 150,
	heatingInterval: 4,
	commandDelay: 5
};
```

**Comment by xseignard** [comment]
Here is a discussion I started a while ago on adafruit forums : http://forums.adafruit.com/viewtopic.php?f=19&p=277966

I think @PaintYourDragon could have some insights since he implemented the python lib. I don't know if the python lib is facing the same behavior

**Comment by xseignard** [comment]
It seems we won't have any insights here.
Closing it for now

**Comment by drskullster** [comment]
Hi @xseignard, any idea why the quickfix from @zettam can't be merged to the code ?

**Comment by xseignard** [comment]
Hi @drskullster, 
Don't hesitate to send a PR and I'll test it ASAP, since I'm working with these thermal printers these days.

Regards,

Xavier

**Comment by drskullster** [comment]
OK I'll do that now.

**Comment by xseignard** [comment]
Great!

**Comment by drskullster** [comment]
I created the PR (#18), I haven't tested it on v0.3.5 (no printer available right now) but it should work.

**Comment by xseignard** [comment]
Merged #18 

Thanks guys! v0.3.6 is available on npm and fixes it.
